samples/binary-search-tree.js

```diff
@@ -1,121 +1,141 @@
-                    // TODO
-                } else {
-                    delete_node.parent.left = delete_node.left;
-                    delete_node.left.parent = delete_node.parent;
-                    delete delete_node.left;
-                    delete delete_node.parent;
-                }
-            } else {
-                let current = delete_node.left;
+			if (!delete_node.left && !delete_node.right) {
+				if (delete_node.parent.right === delete_node)
+					delete delete_node.parent.right;
+				else
+					delete delete_node.parent.left;
+				delete delete_node.parent;
+			} else if (!delete_node.left) {
+				if (delete_node.parent.right === delete_node) {
+					// TODO
+				} else {
+					delete_node.right.parent = delete_node.parent;
+					delete_node.parent.left = delete_node.right;
+					delete delete_node.right;
+					delete delete_node.parent;
+				}
+			} else if (!delete_node.right) {
+				if (delete_node.parent.right === delete_node) {
+					// TODO
+				} else {
+					delete_node.parent.left = delete_node.left;
+					delete_node.left.parent = delete_node.parent;
+					delete delete_node.left;
+					delete delete_node.parent;
+				}
+			} else {
+				let current = delete_node.left;
 
-                while (current.right)
-                    current = current.right;
+				while (current.right)
+					current = current.right;
 
-                if (current === delete_node.left) {
-                    // TODO
-                } else if (current.left) {
-                    // TODO
-                } else {
-                    delete current.parent.right;
-                    delete current.parent;
-                    current.left = delete_node.left;
-                    current.right = delete_node.right;
-                    current.left.parent = current;
-                    current.right.parent = current;
-                    delete delete_node.left;
-                    delete delete_node.right;
-                    this.root = current;
-                }
-            }
-        }
-    }
+				if (current === delete_node.left) {
+					// TODO
+				} else if (current.left) {
+					// TODO
+				} else {
+					delete current.parent.right;
+					delete current.parent;
+					current.left = delete_node.left;
+					current.right = delete_node.right;
+					current.left.parent = current;
+					current.right.parent = current;
+					delete delete_node.left;
+					delete delete_node.right;
+					this.root = current;
+				}
+			}
+		}
+	}
 
-    search(val) {
-        // TODO
-    }
+	/**
+	 * Returns the node containing the specified value.
+	 * @param val the value to be searched
+	 * @returns {Node} containing value or new Node containing error message
+	 */
+	search(val) {
+		if(this.root === null) return new Node("Node has not been found.");
+		if(this.root.val === val) return this.root;
+		let result;
+		if(val < this.root.val && this.root.left){
+			this.root = this.root.left;
```

Comment:
This implementation first overwrites the root field of a BST, and then restores afterwards.  I recommend you to use a local variable to remember the current node of the search.  In fact, the current implementation does not restore the root field when the value is not found (see line 132).

Reply:
Normally I would implement a BST the way you described above, but when I tried to do that, I encountered the following things:

1. Ogushi-kun added a parent field in his node class, which is unusual for a BST tree, but I decided to build upon his work.

2. Because Kanon lacks GC, every new instance of BST() created because of recursion would still be displayed after exiting the function scope in Kanon. 

Nevertheless, I have fixed the bug, in which the root field was not restored if the node was not found.